xarray/core/nanops.py

```diff
@@ -83,8 +99,13 @@ def nanargmax(a, axis=None):
 
 
 def nansum(a, axis=None, dtype=None, out=None, min_count=None):
-    mask = isnull(a)
-    result = np.nansum(a, axis=axis, dtype=dtype)
+    if hasattr(a, "__array_namespace__"):
+        a, mask = _replace_nan(a, 0)
+        xp = get_array_namespace(a)
+        result = xp.sum(a, axis=axis, dtype=dtype)
+    else:
+        mask = isnull(a)
+        result = np.nansum(a, axis=axis, dtype=dtype)
```

Comment:
it seems I spoke too soon... in that thread immediately afterwards this is rejected as not being supported by all libraries (and no further mention, unfortunately).

@shoyer, would you happen to know what we can do here? For `nansum` this is not really critical (at least for `pint` since we're replacing with `0`), but I imagine we want to keep the code as close as possible to `nanprod` (and for that we'd need to replace with a dimensionless `1`, which `pint` will complain about)

Reply:
How about:
```python
masked = where(isnull(a), zeros_like(a), a)
return sum(masked, axis=axis, dtype=dtype)
```

Most of the other nan reductions can also be pretty pretty succinctly from scratch in terms of `where` and `isnull`.

Reply:
Unfortunately due to the way `pint` implements `*_like` (see hgrecco/pint#882) this turns out to be pretty much the same as the implementation of `_replace_nan` (which I removed in #6873). For `sum` this will work because `0` is compatible with every unit, but `1` (for `prod`) is not.

However, we might also call this a bug in `pint`: maybe
```python
ones_like(q)
```
should be the same as
```python
full_like(q, fill_value=Quantity(1, q.units))
```
?

Reply:
yes, that sounds good to me.